### ISSUE TITLE
Strange abm function parameters

### ISSUE THREAD

**Issue by HybridDog** [minetest/minetest#6507]
The parameters of the function in the abm definition are not explained in lua_api.txt.
`action = func(pos, node, active_object_count, active_object_count_wider)`
pos and node are self-explanatory, but active_object_count and active_object_count_wider aren't.
Some time ago I wanted to use these two to detect how many items are next to and near the node, but it didn't work. What are these parameters used for?
I've never seen a mod which uses them btw.

**Comment by SmallJoker** [comment]
`active_object_count` is the count of objects that are stored in the current mapblock of the triggered node.
`active_object_count_wider` is the count of objects in the surrounding mapblocks, within a cubic radius of 1.
Explanation found in the source code, `serverenvironment.cpp`.

**Comment by HybridDog** [comment]
Does it include unloaded objects?

**Comment by paramat** [comment]
I'd like these documented also, they've always been a mystery to me.

**Comment by HybridDog** [comment]
The meaning of day and night in vmanip:set_lighting also isn't documented: https://github.com/minetest/minetest/blob/master/doc/lua_api.txt#L3960

**Comment by HybridDog** [comment]
‫

**Comment by HybridDog** [comment]
I've printed those two abm function parameters. active_object_count doesn't seem to be the number of items in the mapblock, it was 1 while active_object_count_wider was two although the two objects (player and dropped item) were at the same place (tested at multiple places). When I dropped the item into the node, active_object_count was 2. Looks like players are detected in the mapblock, whereas luaentities are detected only inside the node.

**Comment by paramat** [comment]
'day' and 'night' seem obvious.

**Comment by HybridDog** [comment]
I don't understand how and for what reason you need to use set_lighting before calc_lighting.

**Comment by numberZero** [comment]
> active_object_count and active_object_count_wider

Are these ever actually used?

> I don't understand how and for what reason you need to use set_lighting before calc_lighting.

To break things. See [screenshot](https://user-images.githubusercontent.com/7352626/32454686-03866d80-c331-11e7-8d1b-747132ecede7.png) =)

**Comment by paramat** [comment]
> I don't understand how and for what reason you need to use set_lighting before calc_lighting.

That's essential for any voxelmanip mod that alters pre-existing world, when you need to clear out the existing light values and set them to zero before recalculating light.

**Comment by numberZero** [comment]
Er, I thought `calc_lighting` does that itself.

**Comment by paramat** [comment]
No, in my early Lua mapgens (that add to core mapgen) i used calc lighting alone and experienced lighting bugs, hmmmm advised i flush out the previous light values with 'set lighting'.
'calc lighting' is designed to work with a completely dark volume of world, this is so it can be used alone when the 'nolight' flag is set which disables a core mapgen's own lighting calculation.

When a mod alters world (either a core mapgen or 'singlenode' empty world) in 'on generated' using the mapgen object voxelmanip, often we can set the 'nolight' flag so that core mapgen is run, then the mod is run and makes it's changes to the world, then finally 'calc lighting' is run once after all that.

However sometimes it's necessary to allow the core mapgen's own lighting calculation to be run, then the mod makes its changes, then light is reset to zero with 'set lighting', then 'calc lighting' is run to re-calculate light.

However we do now have
```
* `write_to_map([light])`: Writes the data loaded from the `VoxelManip` back to the map.
    * if `light` is true, then lighting is automatically recalculated.
      The default value is true.
```
But this automatic light recalculation only occurs for non-mapgen object voxelamnips.

**Comment by HybridDog** [comment]
What's the reason for not automatically setting the light values to 0 if calc_lighting is called while the nolight flag isn't set?

Also, when there are many mods which need light recalculation (invoke calc_light), why don't you simply recalculate light once after the all the on_generates are executed instead of calculating it again and again, for each calc_light invocation?

**Comment by paramat** [comment]
I guess because we already have 'set lighting', so it's simpler to use that than add code to 'calc lighting' that does the same.

> when there are many mods which need light recalculation (invoke calc_light), why don't you simply recalculate light once

Ideally yes, but if each mod is written to do 'calc lighting' then we can't really alter the code of the mod because a mod should always run as it is coded.

**Comment by HybridDog** [comment]
You could add a request_light_recalculation function which does this and make calc_light and set_light deprecated in lua_api.txt (but still available for compatibility and special purpose).

**Comment by paramat** [comment]
#6906 merged.
